Deprecate implicit conversion of Duration

skipkayhil opened the pull request:
The presence of method_missing on Duration can lead to some very confusing behavior:

```ruby
1.year.days # => 31556952 Days
```

Methods to convert from duration to a specific unit were added in 7bd9603778c059d8f0969feaadb26672534dbece and a2535d9fe91f58aafc7206a8a51dd29a06ecbf61, but the confusing methods they were meant to replace were left in.

This change fixes this by deprecating method_missing and reimplementing or explicitly delegating methods that make sense to keep.

In addition, there are many existing methods on Duration that can result in surprising results or incorrect units. Many methods will treat Numerics implicitly as a number of seconds, which can easily lead to unexpected behavior. In addition, there are cases like two Durations multiplied together which should result in a unit of time^2 and so should not return another Duration. These specific instances have also been deprecated:

- equality of Duration and Numeric (`1.minute == 1 # => false`)

- comparing Durations and Numerics (`5 > 2.minutes # => false`)

- adding/subtracting Durations and Numerics (`5 + 2.minutes # => 2
  minutes and 5 seconds`)

- Numeric / Duration (`3 / 2.seconds # => 1` but should have a unit of `second^-1`)

- Numeric % Duration (`3 % 2.minutes # => 3 seconds`)

- Duration * Duration (`3.days * 2.days # => 518400 days` but should really be `6 days^2`)

Other deprecations:

- Duration#is_a? returning true for value's class

- Duration#instance_of? returning true for value's class

Fixes #45433

skipkayhil committed:
Deprecate implicit conversion of Duration

The presence of method_missing on Duration can lead to some very
confusing behavior:

```ruby
1.year.days # => 31556952 Days
```

Methods to convert from duration to a specific unit were added in
7bd9603 and a2535d9, but the confusing methods they were meant to
replace were left in.

This change fixes this by deprecating method_missing and reimplementing
or explicitly delegating methods that make sense to keep.

In addition, there are many existing methods on Duration that can result
in surprising results or incorrect units. Many methods will treat Numerics
implicitly as a number of seconds, which can easily lead to unexpected
behavior. In addition, there are cases like two Durations multiplied
together which should result in a unit of time^2 and so should not
return another Duration. These specific instances have also been
deprecated:

- equality of Duration and Numeric (`1.minute == 1 # => false`)

- comparing Durations and Numerics (`5 > 2.minutes # => false`)

- adding/subtracting Durations and Numerics (`5 + 2.minutes # => 2
  minutes and 5 seconds`)

- Numeric / Duration (`3 / 2.seconds # => 1` but should
  have a unit of `second^-1`)

- Numeric % Duration (`3 % 2.minutes # => 3 seconds`)

- Duration * Duration (`3.days * 2.days # => 518400 days` but should
  really be `6 days^2`)

Other deprecations:

- Duration#is_a? returning true for value's class

- Duration#instance_of? returning true for value's class

matthewd commented:
These methods are the most egregiously confusing, but I wonder if we should go further, and deprecate the `method_missing` entirely, carving out an explicit delegation for any specific methods we think are worth keeping. I'm inclined to claim that any time a duration silently acts like an integer is similarly surprising.

matthewd commented:
`abs` and `zero?` both sound reasonable to me too: a duration isn't an integer, but is (loosely, though not literally) numeric, so it has the concept of a "zero" value, and of being positive/negative.

That `times` test just sounds like it's an arbitrary choice of delegated method, not an affirmative expectation that a duration is countable (which goes to the idea that it's the bad sort of method: unlike the above that [can] treat the duration as a self-defining quantity, `times` is treating it as a count of seconds.

Given it's always been a full delegation, we'll probably need to go through the full list of methods on Integer... there's no historical reason this class would have comprehensive test coverage of its API.

natematykiewicz commented:
I spotted this like a month ago and should have done something about it. I don't understand the purpose of the method_missing. I feel like it only causes problems.

https://twitter.com/natematykiewicz/status/1529710207059779586

natematykiewicz commented:
Also, I was never able to track down when it was added. All I know is it's at least 15 years old and was added when the repo was SVN. Here's the oldest git commit I could find with the method missing in it.

https://github.com/rails/rails/commit/887870f20c347179aef0545ee2019c02ed9f74d1

skipkayhil commented:
@matthewd updated to implement all of the methods we discussed, I still need to write more tests and documentation but figured I'd ask for another round of feedback first

> Also, I was never able to track down when it was added. All I know is it's at least 15 years old and was added when the repo was SVN. Here's the oldest git commit I could find with the method missing in it.
> 
> [887870f](https://github.com/rails/rails/commit/887870f20c347179aef0545ee2019c02ed9f74d1)

@natematykiewicz  It looks to be part of the original implementation when `Duration` was added in 276c9f29cde80fafa23814b0039f67504255e0fd. My understanding is that since `1.days` and friends previously returned integers, `method_missing` was chosen at the time to reduce the friction of that change.

natematykiewicz commented:
@skipkayhil Ahh, that makes sense. But nowadays there's no real expectation that it's numeric, so the `method_missing` can do a deprecation cycle and eventually get removed. Nice!

natematykiewicz commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -312,6 +319,63 @@ def %(other)
+    end
+
+    def divmod(other)
+      [div(other), self % other)]
```
Buildkite is failing. This either needs another left parenthesis or the right one removed.

```suggestion
      [div(other), self % other]
```

natematykiewicz replied:
I also can't help but wonder if this would be more efficient written as `self.to_i.divmod(other.to_i)`. Then you'd be using a C function.

skipkayhil replied:
> Buildkite is failing. This either needs another left parenthesis or the right one removed.

Ack, thank you

> I also can't help but wonder if this would be more efficient written as `value.divmod(other.value)`. Then you'd be using a C function.

I'm not sure that's the same thing as this. The delegation will return something like `[Numeric, Numeric]` whereas how I have it currently will be `[Duration, Duration]` depending on the type of `other`.

natematykiewicz replied:
> Ack, thank you

No problem!

> I'm not sure that's the same thing as this. The delegation will return something like `[Numeric, Numeric]` whereas how I have it currently will be `[Duration, Duration]` depending on the type of `other`.

Ahh, I overlooked that. And both the `div` and `%` functions on Duration already use the Numeric ones. Cool. Nevermind me.

matthewd commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -371,6 +442,7 @@ def to_s
     def to_i
       @value.to_i
     end
+    alias :to_int :to_i
```
This one is tricky... it could be argued either way for an integer-based duration, but I think it's definitely wrong (and not currently there) for e.g. a float-based one.

Maybe we just drop it, and let the deprecation push people to an explicit conversion?

matthewd commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -46,6 +47,11 @@ def test_to_s
 
+  def test_in_milliseconds
+    assert_equal 86400000, 1.day.in_milliseconds
+    assert_equal 60000, 1.minute.in_milliseconds
```
`in_seconds` is arguably the one that's [already] wrong and at fault here, but:

```suggestion
    assert_equal 60000, 1.minute.in_milliseconds
    assert_equal 1500, 1.5.seconds.in_milliseconds
```

skipkayhil replied:
Hmm, maybe `in_seconds` should just return `@value`? Especially if we're going to keep `to_f`, `to_r`, etc.

matthewd commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -763,6 +766,71 @@ def test_duration_symmetry
+
+    methods.each do |method|
+      assert_deprecated do
+        1.year.public_send(method) rescue ArgumentError
```
That `rescue` syntax isn't doing what it appears

skipkayhil replied:
Hmm, I had added it because some of the deprecated methods take an argument and this seemed to quiet those errors. Am I missing something about how that works? (I figured the `assert_deprecated` is ensuring that the deprecation is getting printed whether or not the delegated method call succeeds)

skipkayhil replied:
Thank you for the explanation!

matthewd commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -312,6 +319,63 @@ def %(other)
         raise_type_error(other)
```
⏫ I think `*(Duration)` is silently treating `other` as Just A Number, so should probably be deprecated too.

matthewd commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -3,6 +3,7 @@
 require_relative "../abstract_unit"
 require "active_support/inflector"
 require "active_support/time"
+require "active_support/core_ext/numeric/bytes" # TODO: remove when Duration#method_missing is removed
```
Much as I don't like having orphan `require`s around the place, the comment is rather unsightly in its own right. How bad would it be for us to push this inside the relevant test method? :thinking:

skipkayhil replied:
Updated this to require in the test and restore methods after. It looks pretty ugly thought, it's certainly making me lean towards just not testing those methods explicitly.

skipkayhil commented:
Finally got back around to this, updated to add more deprecations and tests for the new methods.

> I think `*(Duration)` is silently treating `other` as Just A Number, so should probably be deprecated too.

Based on this comment, I thought more about what other operations don't make sense without implicit conversion and added deprecations for the following: 
- comparing Durations and Numerics (`5 > 2.seconds`)
- adding/subtracting Durations and Numerics (`5 + 2.seconds`)
- dividing a Numeric by a Duration (`3 / 2.seconds`)
- Numeric % Duration (`3 % 2.seconds`)
- Duration * Duration (`3.seconds * 2.seconds`)
- `Duration#is_a?` returning true for `value`'s class
- `Duration#instance_of?` returning true for `value`'s class
- equality of Duration and Numeric (`1.minute == 60`)

Something I'm not entirely sure what to do with is `%` and related methods (`modulo`, `divmod`). `%` currently always implicitly treats the caller as a number of seconds:

```ruby
5.minutes % 120 # => 1.minute
7.seconds % 3.seconds # => 1.second

7.minutes % 3 # => 0 
# because 420 seconds is divisible by 3
```

Should `Duration % Numeric` just be deprecated? Or maybe we try to fix it so `7.minutes % 3 # => 1.minute`? I've thought about this for a bit but definitely need to think some more.
